posthog/api/test/test_retention.py

```diff
@@ -0,0 +1,165 @@
+    database in the right state
+    """
+    person = Person.objects.create(team_id=team_id)
+    PersonDistinctId.objects.create(distinct_id=distinct_id, team_id=team_id, person_id=person.id)
+    capture_event(
+        event=EventData(
+            event="$identify",
+            team_id=team_id,
+            distinct_id=distinct_id,
+            timestamp=datetime.now(),
+            properties={"distinct_id": distinct_id},
+        )
+    )
+
+
+def get_retention(
+    client: Client,
+    display: str,
+    events: List[Dict[str, Any]],
+    date_from: str,
+    date_to: str,
+    selected_interval: int,
+    total_intervals: int,
+    target_entity: Dict[str, Any],
+    returning_entity: Dict[str, Any],
+    insight: str,
+    period: str,
+    retention_type: str,
+    properties: Optional[List[Dict[str, Any]]] = None,
+    filter_test_accounts: bool = True,
+):
+    return client.get(
+        "/api/person/retention",
+        data={
+            "display": display,
+            "events": json.dumps(events),
+            "date_from": date_from,
+            "date_to": date_to,
+            "selected_interval": selected_interval,
+            "total_intervals": total_intervals,
+            "filter_test_accounts": filter_test_accounts,
+            "insight": insight,
+            "period": period,
+            "retention_type": retention_type,
+            "target_entity": json.dumps(target_entity),
+            "returning_entity": json.dumps(returning_entity),
+            "properties": json.dumps(properties or []),
+        },
+    )
+
+
+@pytest.mark.django_db
+@freeze_time("2021-08-03")
+def test_insight_retention_missing_persons_gh_5443(client: Client):
+    """
+    This is a regression test for GH-5443.
+
+    The scenario here is that, an api request is being made for person retention, specifically for:
+
+      1. a "Week" period is being requested
+      2. events just over a week from the first event for a user
+
+    """
+
+    organization = create_organization(name="test org")
+    team = create_team(organization=organization)
+    user = create_user("user", "pass", organization)
+
+    identify(distinct_id="abc", team_id=team.id)
+
+    #  This event will be the first event for the Person wrt the retention
+    #  period
+    capture_event(
+        event=EventData(
+            event="event_name", team_id=team.id, distinct_id="abc", timestamp=datetime(2021, 3, 29), properties={},
+        )
+    )
+
+    # Create an event for just over a week from the initial identify event
+    capture_event(
+        event=EventData(
+            event="event_name", team_id=team.id, distinct_id="abc", timestamp=datetime(2021, 4, 5), properties={},
+        )
+    )
+
+    client.force_login(user)
+
+    # These params are taken from
+    # https://sentry.io/organizations/posthog/issues/2516393859/events/df790b8837a54051a140aa1fee51adfc/?project=1899813
+    response = get_retention(
```

Comment:
is there a benefit to putting this test in api/test and testing an entire api call rather than queries/test? asking because this strays away from the current pattern we have for query related tests which is fine just curious if you're thinking we should modify all tests to follow this pattern more

Reply:
My thinking was that:

* this should be a pretty stable test in terms of if we want to change the internal implementation
* I also wouldn't be sure as to what level internally we'd want to test, where as at this rest api level it's pretty obvious
* when creating a regression test, it's pretty handy to be able to pretty much copy the request as logged in sentry. Potentially just copying that curl command from sentry would be even easier.

Maybe if this query is shared elsewhere then we're want to test at a different level to test as a unit in its own right.  I would probably do the same again for other sentry bugs. I would have a preference for tests at this level which capture the interaction with the system close to real world interaction with the API, which should give us a bit more flexibility to move things around internally.